File: 240206/부분 수열의 합/the-sum-of-the-subsequences.py
n, m = map(int, input().split())
arr = [0]+list(map(int, input().split()))
dp = [
    [False]*(m+1)
    for _ in range(n+1)
]
dp[0][0] = True

for i in range(1, n+1):
    for j in range(1, m+1):
        if j>=arr[i] and dp[i-1][j-arr[i]]:
            dp[i][j] = True
        elif dp[i-1][j]:
            dp[i][j] = True
print('Yes') if dp[n][m] else print('No')

File: 260827/부분 수열의 합/the-sum-of-the-subsequences.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        int[] arr = new int[n];
        boolean[] dp = new boolean[m+1];
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }
        dp[0] = true;
        for(int i = 0; i < n; i++){
            for(int j = m; j >= 1; j--){
                if(j >= arr[i] && dp[j-arr[i]]){
                    dp[j] = true;
                }
            }
        }
        System.out.print(dp[m] ? "Yes" : "No");
    }
}